### PR TITLE
[ENH] Added support detection for number of clusters, helping with hyperparameter tuning

### LINKED ISSUE #7781: [ENH] Support detection for number of clusters
In a non-time-series situation, identification of optimal number of clusters is quite common, and people usually do elbow curves with SSW, maximum silhouette score, etc. It will be good to have such capability in `sktime` to find the optimal number of groups while performing time series clustering.

1. The user must be able to specify a metric of choice, e.g. intertia (available as model attributes), silhouette score (https://tslearn.readthedocs.io/en/stable/gen_modules/clustering/tslearn.clustering.silhouette_score.html), etc.
2. User may choose to pass an array of number of clusters to try.
3. Alternatively, user may specify minimum and maximum number of clusters.

### PR DESCRIPTION
**Pull request by TomatoChocolate12**
#### Reference Issues/PRs

Resolves #7781 .

#### What does this implement/fix? Explain your changes.

I have tried to implement a class called `ClusterSupportDetection` within a new file called `_tune.py` to help with hyperparameter tuning with clustering methods. Two support detection techniques, using *silhouette score* and the *elbow method* have been implemented. Some liberty has been taken with the implementation of the elbow method technique, wherein, the user can choose to use a *threshold* to ensure that the elbow chosen (optimal number of clusters) would have a slope lesser than the specified threshold.

#### Does your contribution introduce a new dependency? If yes, which one?

I have used the external libraries of `numpy` and `sklearn.metric` for this task. I do not think these would cause any dependancy issues.

#### What should a reviewer concentrate their feedback on?

E

#### Did you add any tests for the change?

Yes, I have added tests with some of the previously implemented clusterers.

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [ ] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

**Review by fkiraly — CHANGES_REQUESTED** [review]
Nice! Great contribution!

* most significant issue is that `X_inner_mtype` is not a panel mtype, and the `sklearn` silhouette score will not work for `sktime`, because the one for time series will expect a `Panel` of time series
* in particular, you will have to design a silhouette score for `sktime`, for time series. I would advise that you subclass `BaseObject` for a `BaseClusterMetric`, which can then use a pairwise panel transformer (e.g., time series metric)  as a component.
* the `metric_params` parameter violates the API requirement that init params are written to self and never changed
* the docstring should say that (a) clusterers need to have the predict capability, and (b) the `inertia_` parameter when the metric is not passed
* `predict` is actually only needed if the metric requires it.

**Comment by yarnabrina** [comment]
Probably this we can use: https://tslearn.readthedocs.io/en/stable/gen_modules/clustering/tslearn.clustering.silhouette_score.html

**Comment by fkiraly** [comment]
> Probably this we can use: https://tslearn.readthedocs.io/en/stable/gen_modules/clustering/tslearn.clustering.silhouette_score.html

in theory yes, but I think this has disadvantages above a plain reimplementation:

* we would need an adapter for the score itself, as it is not a proper estimator, so will "cost" as much work as a reimplementation
* the `tslearn` interface allows only equal length data
* the `tslearn` interface allows only metrics in the form of callables, it would be better if this could *also* be `sktime` distances - becaues then we can tune parameters of the distances inside the silhouette score!
* we get a dependency on `tslearn` for a basic tuning and evaluation task in clustering

Given that I estimate the cost in time and work as equal, with the only disadvantage of the new implementation that we have to design it, I would go with a new implementation.

**Comment by TomatoChocolate12** [comment]
again sorry for the delay in response and work @fkiraly and @yarnabrina . i was caught up with some projects in my university and was not able to dedicate time to this.

i have made the suggested changes and have taken inspiration from tuners for other methods to do the same. i actually named the file `_tune.py` initially because the tuner for the forecasting module is also named `_tune.py`. i have changed the name now to make it a public module. i have also taken care of the other small changes and formatted the code a little bit: hope this satisfies all the tests.

please take a look at it and drop a review to let me know if i need to change some other things. thanks!

File: sktime/clustering/tune.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+    def _fit(self, X, y=None):
+        """Finds the optimal number of clusters by evaluating different values."""
+        best_score = -np.inf
+        best_param = None
+        scores = {}
+        inertia_values = []
+
+        for param in self.param_range:
+            if self.verbose:
+                print(f"Evaluating {param} clusters...")
```
Comment (yarnabrina) [inline sktime/clustering/tune.py:171]: Should we use print, or logging? Thoughts?

File: sktime/clustering/tune.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+                inertia = estimator.inertia_
+                inertia_values.append(inertia)
+                scores[param] = -inertia  # Minimizing inertia
+
+        if not self.metric:
+            # Find the elbow point
+            best_param = self._find_elbow_point(self.param_range, inertia_values)
+            best_score = -inertia_values[best_param - self.param_range[0]]
+        else:
+            best_param = max(scores, key=scores.get)
```
Comment (yarnabrina) [inline sktime/clustering/tune.py:193]: Question: is it ensured that all metrics need to be maximised? Should direction be configurable?

File: sktime/clustering/tune.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+
+        if self.verbose:
+            print(f"Best parameter: {best_param} with score: {best_score}")
+
+        return self
+
+    def _find_elbow_point(self, param_range, inertia_values):
+        """Finds the elbow point by detecting when the slope drops below a threshold."""
+        diffs = np.diff(inertia_values)
+        threshold = self.metric_params.get("elbow_threshold", 0.1)
```
Comment (yarnabrina) [inline sktime/clustering/tune.py:208]: If you are expecting a specific parameter name, it should be documented explicitly.

The default threshold value is slightly odd. Will it not make sense to do it with respect to second derivatives?

(Currently I simply use `kneed`.)

**Review by yarnabrina — COMMENTED** [review]
Added few comments

File: sktime/clustering/tune.py (hunk 1 above)
Comment (TomatoChocolate12) [reply sktime/clustering/tune.py:171]: It would make sense to just log the statement as there is a chance `param_range` is quite large and that would result in an unhealthy amount of printed lines in the terminal. I'll make this change.

File: sktime/clustering/tune.py (hunk 2 above)
Comment (TomatoChocolate12) [reply sktime/clustering/tune.py:193]: as silhouette score and elbow point have been implemented, i did not consider the idea of making the direction configurable. i'll take a look at some other cluster support evaluation metrics and see if direction should be made configurable. but this cannot be an argument passed to the class right?

File: sktime/clustering/tune.py (hunk 3 above)
Comment (TomatoChocolate12) [reply sktime/clustering/tune.py:208]: What I'm doing here is trying to find the point where the slope becomes closest to being horizontal. If I were to use the second derivative, I would need to identify the point where the slope changes the most, right? However, my intuition is that this approach could lead to misleading results. For instance, a very small cluster size (on the x-axis) might show a large change in slope compared to the previous point, but that wouldn't necessarily indicate the optimal cluster size.

Given the inherent subjectivity of the elbow method, finding a definitive answer is quite challenging. That said, I’ll ensure the parameter is well-documented for clarity. Thanks for the review! :)

File: sktime/clustering/tune.py (hunk 2 above)
Comment (TomatoChocolate12) [reply sktime/clustering/tune.py]: I have made direction configurable.

File: sktime/clustering/tune.py [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+    get default parameters for test instance(s) - get_test_params()
+
+"""
+
+from sktime.base import BaseObject
+from sktime.param_est.base import BaseParamFitter
+from sktime.distances import pairwise_distance
+import numpy as np
+
+class BaseClusterMetric(BaseObject):
```
[the remark this answers is not in the document]
Comment (TomatoChocolate12) [reply sktime/clustering/tune.py]: Done.

File: sktime/clustering/tune.py (hunk 3 above)
Comment (TomatoChocolate12) [reply sktime/clustering/tune.py]: I have not made any changes in my most recent commit regarding this, but I would be happy to hear your thoughts on the reason that I gave. Mostly because I am myself not fully convinced. (tagging @fkiraly as their input would be quite valuable as well)

File: sktime/clustering/tune.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (yarnabrina) [reply sktime/clustering/tune.py]: Sorry for the miscommunication.

What I meant is where you are defining the base metric class and then inheriting it for silhouette, that's bettrr suited in `performance_metrics`. The tuning part is of course going to be part of the `clustering`.

File: sktime/clustering/tune.py (hunk 3 above)
Comment (yarnabrina) [reply sktime/clustering/tune.py]: (I've been out of touch of academics for so long, so apologies for some naive comments below.)

If you are going for the slope idea, I am not sure if the threshold of 0.1 on diff is enough, since it's likely going to be affected by scale of data. Also, probably to be safe about horizontal inference, we need to see consistent low slopes. I am not sure if checking by one diff (because the function returns at first find) will be enough.

I understand your point of initial high drop in second differences. I'm not sure how to answer that, so let me think for a while.

File: sktime/clustering/tune.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (TomatoChocolate12) [reply sktime/clustering/tune.py]: made the changes.

File: sktime/clustering/tune.py (hunk 3 above)
Comment (TomatoChocolate12) [reply sktime/clustering/tune.py:149]: yeah, i agree. i think it would be a good idea to use the second *and* first differential for this task. i think both of them neatly close each other's faults. what do you think?

it would ensure that the elbow would be a somewhat horizontal slope, and at the same time, also ensure values are small enough as there is a bigger change in slope? but i dont know how to integrate them together.

**Comment by TomatoChocolate12** [comment]
@yarnabrina i looked at the `kneed` library that you suggested and took a look at the kneedle algorithm. they have published a paper where they were trying to find the point of maximum curvature for the elbow point. the mathematical relation for this seemed to utilise both the first and second derivatives like i wanted to, so i decided to implement this. please take a look at it. thanks.

**Comment by TomatoChocolate12** [comment]
also this method does not even require a threshold of sorts so even more ambiguity is removed.

**Comment by yarnabrina** [comment]
Looks fine to me logic wise, but may need minor changes just for integrating with the codebase (not changing init params, doc export, etc.). Unfortunately I'm a bit busy this week and the next, so unable to do a thorough review right now. Hopefully other reviewers will be able to help you sooner.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Nice contribution!

For API compatibility, ensure:

* that the code is formatted per the pre-commit https://www.sktime.net/en/stable/developer_guide/coding_standards.html
* that `__init__` params are written to self as passed, but never changed after that.

See here for the adding estimators guide: https://www.sktime.net/en/latest/developer_guide/add_estimators.html

**Comment by fkiraly** [comment]
Great, I restarted the tests now. Formatting is all fine. If you ran `check_estimator` and all passed, then the tests should also pass.

**Review by fkiraly — CHANGES_REQUESTED** [review]
The error is due to soft dependencies required in the estimator passed to `get_test_params`.

This is not a soft dependency of the parameter fitter itself. Hence, is there another estimator without soft dependencies we could use here?

**Comment by TomatoChocolate12** [comment]
are libraries like `numpy` and `collections` also soft dependencies because of which the test is failing? if that is the case, can i use libraries estimators which use libraries like `sklearn`?

**Comment by TomatoChocolate12** [comment]
I think the estimator `TimeSeriesKvisibility` does not require any extra dependencies. Please run the tests again and hopefully it passes all the checks this time around.

**Comment by TomatoChocolate12** [comment]
i have figured out the fix for the soft dependency issue, but i was stuck with trying to resolve testing issues with `check_estimator`. the code that i have committed has two added functions within the `ClusterSupportDetection` class, `get_params` and `set_params`. these functions are defined for a super-super-super class of `ClusterSupportDetection` [`BaseObject`]. but, as the tests corresponding to these functionalities were not passing, I needed some help.

**Review by fkiraly — CHANGES_REQUESTED** [review]
I think this is great - the metric design makes sense, later we can add the inner layer.

Some API remarks:

* the `metric` passed is expected to be an object instance, not a class, so we do not need to pass the `metric_params` separately. Assume `metric` is an instance.
* This should also remove the need to implement `get_params` and `set_params`, these should never be overridden.
* For `_get_fitted_params`, I suggest to not ovevride it. Simply name the fitted parameter the same as the attribute. If you want `best_n_clusters` as the name, write to `best_n_clusters_`. Though, should it not be `best_n_clusterers_`?
* For `get_test_params`, you are using a base class, `BaseTimeSeriesLloyds` - I think this is what causes the current test errors.
* I would also suggest to implement a `predict` method, and `predict_proba`. This should delegate to the inner clusterer's `predict` resp `predict_proba`, with best parameters. Of course only if the inner clusterer has this capability. This can be controlled by setting the tags `capability:predict` and `capability:predict_proba` in `__init__`.

**Comment by TomatoChocolate12** [comment]
i have made the requested changes, i got rid of testing with `BaseTimeSeriesLloyds`. I thought it was a clusterer based on the description, but I should have looked more clearly. The tests all seem to be passing for now, please run the workflow again.
